Explicitly define ScreenLogic entity descriptions

dieselrabbit opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Explicitly define entity descriptions for sensor, binary_sensor, and number entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #92475
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

dieselrabbit committed:
Explicitly define sensor, binary_sensor, and number entities

dieselrabbit committed:
Remove path_template test

MartinHjelmare commented on homeassistant/components/screenlogic/binary_sensor.py:
```diff
@@ -104,67 +220,58 @@ async def async_setup_entry(
+            pump_sensor_description = copy(proto_pump_sensor_description)
+            device, group, value = pump_sensor_description.data_path
+            group = int(pnum)
+            pump_sensor_description.data_path = (device, group, value)
```
Move out the parts that are not specific for the description, ie group, unique_id generation and name. Those don't have to be set on the entity description but can be set in the entity.

dieselrabbit replied:
Do you mean apply them as kwargs in the class constructor, or pull them all the way into [ScreenLogicEntity's init](https://github.com/home-assistant/core/blob/75951dd67be7c2e149348c6171a9af5b07d4de8a/homeassistant/components/screenlogic/entity.py#L45-L64)? (which could be possible)

MartinHjelmare replied:
If all the entities share the parameters we can accept them and set the features in the base class init method. Otherwise do it in each platform entity init method.

dieselrabbit replied:
[77394cc](https://github.com/home-assistant/core/pull/100173/commits/77394cc35ab100aba66d6d8ef60c5daea99d6ef0)

MartinHjelmare reviewed (commented):
This is so much easier to read. 👍

bdraco commented:
Awesome. Much cleaner. Thanks @MartinHjelmare and @dieselrabbit

dieselrabbit committed:
Refactor entity setup

dieselrabbit committed:
test_init coverage

dieselrabbit committed:
fix type check

MartinHjelmare commented on homeassistant/components/screenlogic/entity.py:
```diff
@@ -50,10 +54,12 @@ def __init__(
-        self._data_path = self.entity_description.data_path
-        self._data_key = self._data_path[-1]
-        self._attr_unique_id = f"{self.mac}_{self.entity_description.key}"
+        self._data_root = self.entity_description.data_root
```
It doesn't look like `self._data_root` is used anywhere outside this method. It doesn't need to be an instance attribute then.

dieselrabbit replied:
[c7a8c41](https://github.com/home-assistant/core/pull/100173/commits/c7a8c41b964a06a32c33c6cc32565c32d669a2af)

MartinHjelmare reviewed (approved):
Nice! Just the comment above.

dieselrabbit committed:
Remove unneeded instance attribute

bdraco commented:
I'll push this to all my production shortly

bdraco commented:
There are a few sensors that went unavailable after installing this PR. All 5 previously had values before

<img width="342" alt="Screenshot 2023-09-17 at 9 46 14 AM" src="https://github.com/home-assistant/core/assets/663432/50408c86-b1df-4c80-8b7c-f679fccec318">

bdraco reviewed (dismissed):
Same thing on the other one I tested

<img width="455" alt="Screenshot 2023-09-17 at 9 47 45 AM" src="https://github.com/home-assistant/core/assets/663432/dd1a1d8f-55ac-43ab-a5e4-949a5a9d15e1">
<img width="664" alt="Screenshot 2023-09-17 at 9 47 39 AM" src="https://github.com/home-assistant/core/assets/663432/35ee2eef-b722-4dbe-bc69-5db22a11b624">

bdraco commented:
Here is the stack trace in the log:

```
2023-09-17 10:43:22.389 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up screenlogic platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/entity.py", line 97, in entity_data
    return self.gateway.get_data(*self._data_path, strict=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/screenlogicpy/gateway.py", line 259, in get_data
    raise KeyError(f"'{key}' not found in '{keypath}'")
KeyError: "'gpm_now' not found in '('pump', 0, 'gpm_now')'"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 359, in _async_setup_platform
    await asyncio.shield(task)
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/sensor.py", line 255, in async_setup_entry
    ScreenLogicPumpSensor(
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/sensor.py", line 338, in __init__
    super().__init__(coordinator, entity_description)
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/sensor.py", line 304, in __init__
    super().__init__(coordinator, entity_description)
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/entity.py", line 61, in __init__
    self._attr_name = self.entity_data[ATTR.NAME]
                      ^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/screenlogic/entity.py", line 99, in entity_data
    raise HomeAssistantError(f"Data not found: {self._data_path}") from ke
homeassistant.exceptions.HomeAssistantError: Data not found: ('pump', 0, 'gpm_now')
```

dieselrabbit commented:
The sensors you have as unavailable are ones where their name or location in the data `dict` changed for this release (`screenlogicpy=0.9.0`).  This seems to indicate that whatever the method being used to push this to production and test is not updating the  referenced version of the `screenlogicpy` python package.

I've seen this happen before when testing updates to `screenlogicpy`, referenced via a github repo. It seems like it happens when a newer version is installed, but then the system is downgraded to a previous version, and then updated again. After the second "upgrade" the old package "stays around" and continues to be referenced. I've had to resort to force upgrading `screenlogicpy` package through the terminal.  Not positive that's what's happening here, though.

I'm going to test again by rebuilding the dev container on `master` branch, then switching to `sl-ent-desc-2` branch.

bdraco commented:
Looks like the library is the right one unless there is another update?

```
homeassistant:/config# pip3 show screenlogicpy
Name: screenlogicpy
Version: 0.9.0
Summary: Interface for Pentair ScreenLogic connected pool controllers over IP via Python
Home-page: https://github.com/dieselrabbit/screenlogicpy
Author: Kevin Worrel
Author-email: kevinworrel@yahoo.com
License: GPLv3
Location: /usr/local/lib/python3.11/site-packages
Requires: async-timeout
Required-by: 
homeassistant:/config# 
```

```
homeassistant:/usr/src/homeassistant/homeassistant/components/screenlogic# git log -1 -w -- .
commit c7a8c41b964a06a32c33c6cc32565c32d669a2af
Author: Kevin Worrel <37058192+dieselrabbit@users.noreply.github.com>
Date:   Sun Sep 17 08:06:22 2023 +0000

    Remove unneeded instance attribute
```

dieselrabbit commented:
That is the expected version.  
I seemed to remember a different location from `site-packages` that was caching the package(s), but maybe that was only for dev container or github refs. 

I'll let you know what I find from the clean upgrade test.

dieselrabbit commented:
For my test, I:
1. Switched branch to `master`
2. Rebuilt the Dev Container
3. Deleted config/device/entity registries
4. Launched HA, only sun & user entities are present.
5. Installed Pentair Screenlogic integration.
6. All expected entities are present
7. Verified `screenlogicpy` = 0.8.2
8. Shutdown HA
9. Switched branch to `sl-ent-desc-2`
10. Launched HA.
11. **All ScreenLogic entities came back, nothing unavailable.**
12. Verified `screenlogicpy` = 0.9.0

Is there another way I should be testing?

bdraco commented:
Looks ok. 

I might have an undetected merge conflict on my integration branch.  I'll do a full reset of the ScreenLogic code and try again

bdraco commented:
Looks like its a clean merge

```
(venv) bdraco@MacBook-Pro-9 home-assistant % git diff -w dieselrabbit/sl-ent-desc-2 -- homeassistant/components/screenlogic
(venv) bdraco@MacBook-Pro-9 home-assistant % 
```

bdraco commented:
One difference in testing, I went from `dev` to this branch instead of `master` to this branch

dieselrabbit commented:
I can test `dev` to here as well. 
I went from `master` to this branch because that's what the non-dev will experience.

dieselrabbit commented:
I saw no issues going from `dev` to here.

bdraco commented:
I'll do a full reset to `dev`, remove the integration, set it up again , and than pull this PR on time

bdraco commented:
I did that but got the same result. I also tried on a fresh local dev setup and its the same.

I'll try to do some debugging now

<img width="337" alt="Screenshot 2023-09-18 at 10 22 22 AM" src="https://github.com/home-assistant/core/assets/663432/66e5c3f5-d262-43c4-8593-7c81c4eff4ac">

bdraco commented:
The data for `gpm_now` is actually missing

```
  'pump': {0: {'data': 134,
               'preset': {0: {'device_id': 6, 'is_rpm': 1, 'setpoint': 1050},
                          1: {'device_id': 1, 'is_rpm': 1, 'setpoint': 1850},
                          2: {'device_id': 2, 'is_rpm': 1, 'setpoint': 1500},
                          3: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          4: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          5: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          6: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          7: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000}},
               'rpm_now': {'name': 'Pool Pump RPM Now',
                           'state_type': 'measurement',
                           'unit': 'rpm',
                           'value': 1050},
               'state': {'name': 'Pool Pump', 'value': 1},
               'type': 2,
               'unknown_at_offset_16': 0,
               'unknown_at_offset_24': 255,
               'watts_now': {'device_type': 'power',
                             'name': 'Pool Pump Watts Now',
                             'state_type': 'measurement',
                             'unit': 'W',
                             'value': 63}},
           1: {'data': 131,
               'gpm_now': {'name': 'Jets Pump GPM Now',
                           'state_type': 'measurement',
                           'unit': 'gpm',
                           'value': 0},
               'preset': {0: {'device_id': 3, 'is_rpm': 1, 'setpoint': 2970},
                          1: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          2: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          3: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          4: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          5: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          6: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000},
                          7: {'device_id': 0, 'is_rpm': 1, 'setpoint': 1000}},
               'rpm_now': {'name': 'Jets Pump RPM Now',
                           'state_type': 'measurement',
                           'unit': 'rpm',
                           'value': 0},
               'state': {'name': 'Jets Pump', 'value': 0},
               'type': 2,
               'unknown_at_offset_16': 0,
               'unknown_at_offset_24': 255,
               'watts_now': {'device_type': 'power',
                             'name': 'Jets Pump Watts Now',
                             'state_type': 'measurement',
                             'unit': 'W',
                             'value': 0}},
```

dieselrabbit commented:
Thank you. I know what the problem is now.

Do you mind if I add one of your previous diagnostic dumps to the test fixtures?

bdraco commented:
> Thank you. I know what the problem is now.
> 
> Do you mind if I add one of your previous diagnostic dumps to the test fixtures?

Feel free to add any information I've posted to github previously to any of the tests.  Its already public 😄

bdraco commented:
[config_entry-screenlogic-1b91af1102c90436d0d01089f35bad9e.json.txt](https://github.com/home-assistant/core/files/12650241/config_entry-screenlogic-1b91af1102c90436d0d01089f35bad9e.json.txt)


Here is the latest in case its useful

dieselrabbit committed:
Skip pump sensors missing data

dieselrabbit committed:
Test platform setup

dieselrabbit commented:
I don't know if the added test is the right way to go about it, but the exception is buried deep in setup and I couldn't find a way to catch it from `hass.config_entries.async_setup()`.  That said, the issue that is causing your missing sensors was fixed in `screenlogicpy` v0.9.1. Don't know if it's better to just bump the api version. 
Either way this change adds a little more protection from missing data.  Open to any suggestions.

MartinHjelmare commented on tests/components/screenlogic/test_init.py:
```diff
@@ -241,3 +251,43 @@ async def test_entity_migration_data(
+            hass=hass,
+        )
+        hass.data.setdefault(DOMAIN, {})[mock_config_entry.entry_id] = mock_coordinator
+        for platform in PLATFORMS:
```
Don't access integration details like hass.data or platforms in the tests. Set up the integration via the config entry while patching the library appropriately.

dieselrabbit replied:
What I was trying to test for was if the platform setup correctly handled missing data for the expected sensor(s). The thought was to eventually use the JSON fixtures to test against a variety of different equipment configurations the integration could encounter.

The problem I was running into with this test in particular was that I couldn't figure out a way catch the exception from as far up as is typically desired by setting up the integration via the config entry.

The exception seemed to be caught in `EntityPlatform._async_setup_platform()` and converted to returning `False`, returned up through `EntityComponent` to `ConfigEntry.async_setup()` where a false result for `await component.async_setup_entry(hass, self)` seems to only be used to set the `state` on the config entry to `ConfigEntryState.SETUP_ERROR`.  I tried asserting on the config entry's state, but it was always 'loaded'.

The only thing I could think to do to catch the exception was to mock/bootstrap the integration's `async_setup()` (what you see with accessing hass.data) and directly call the integration's platform `async_setup_entry()`s.

I'm sure there are reasons, but it is a little frustrating to seemingly not be able to assert the success of platform setup from `hass.config_entries.async_forward_entry_setups()`. Maybe there's an avenue that I'm missing?

MartinHjelmare replied:
I'm not following about the exception to catch. There should be no exception, right? We can assert that the expected entities have created a state in the state machine.

dieselrabbit replied:
If the platform setup correctly skips setting up entities where there will be missing data, then correct there should be no exception. 

I was trying to make the test more (maybe too much) generic in that it could catch missing data that we didn't think about by catching the exception that is raised when the expected data is not found. Sort of a "Canary in the coal mine". But you're right; no exception != good test. In this case the test could pass but the validity of the entities is still uncertain. We should be asserting on the actual expected state of things.  It was just that there are so many permutations of that to test, and that are not currently known.

This all may be moot, though. The most recent version of the api library doesn't have any data excluded so we shouldn't run into this anyways. Still interested in expanding my knowledge of HA though.

MartinHjelmare replied:
If there's an exception during platform setup no entities of that platform will be added. So it's enough to check that one entity per platform is added to know that platform setup was successful while creating the missing data case for each platform by patching the library data.

dieselrabbit replied:
b7a2cbd

bdraco commented:
> That said, the issue that is causing your missing sensors was fixed in `screenlogicpy` v0.9.1. Don't know if it's better to just bump the api version.

If its compatible, I'd do a separate PR to bump the api and than rebase this one after we merge it

dieselrabbit commented:
Should I back out the bad test or refine it?

dieselrabbit committed:
Refactor test_platform_setup

dieselrabbit committed:
Merge branch 'dev' into sl-ent-desc-2

dieselrabbit committed:
Merge branch 'dev' into sl-ent-desc-2

MartinHjelmare commented on tests/components/screenlogic/test_init.py:
```diff
@@ -272,22 +273,26 @@ def add_ents(ents: list[Any]) -> None:
-                    f"homeassistant.components.screenlogic.{platform}"
-                ),
+        if EQUIPMENT_FLAG.CHLORINATOR in gateway.equipment_flags:
+            assert (
```
Don't use any conditions in tests. We should know how the gateway is configured since we set up the test.

dieselrabbit replied:
I was using parameterize to test against different gateway configurations, but as this test is really supposed to only be testing setup with missing data, I can build a single data set to support that.

MartinHjelmare replied:
We can extract the entity_ids that we want to assert state for, into another parameter. Then do a for loop here over those entity_ids.

dieselrabbit replied:
6d7eae3

dieselrabbit committed:
Simplify test_platform_setup

dieselrabbit committed:
Merge branch 'sl-ent-desc-2' of https://github.com/dieselrabbit/core into sl-ent-desc-2

dieselrabbit committed:
Merge branch 'dev' into sl-ent-desc-2

MartinHjelmare reviewed (approved):
Thanks!

bdraco commented:
Will retest shortly and merge if all good

bdraco reviewed (approved):
Thanks @dieselrabbit 
Thanks @MartinHjelmare 

I know this one was a lot of time investment on everyone's part. Its appreciated

bdraco commented:
Tested and working on 2 production systems

dieselrabbit commented:
Thank you both for your patience with me!
